### PR TITLE
feat: full blocks, bundling & compression, blob usage optimization & tracking

### PR DESCRIPTION
**Pull request by segfault-magnet**
closes: #116
closes: #69 
closes: #124 (thanks to @hal3e for contributing)
A big thank you to @MujkicA for all the research that went into this.

### TLDR;
The committer will now stream full blocks from the fuel network, try to find an optimal way to bundle them so that we maximize value for money on L1 and proceed to post the bundle split into blobs to Ethereum.

Preliminary results on testnet fuel block data show that, if given enough time to optimize, we are consistently achieving a 3.0 compression ratio (5.0 + without block headers) and blob utilization of +96% (3% of that is [blob encoding overhead](https://github.com/FuelLabs/fuel-block-committer/issues/108), the rest is our compressed data). (Ran a test during the CEST day for about 3-4 hours with `blocks_to_accumulate` of `3600`)

To bundle 1h worth of fuel blocks (around 3600 blocks) optimally takes around 20s for my setup (ryzen 9 7950x). We'll probably bump that time limit to 3-5 minutes in prod. 

### How it works
`block_height_lookback_window` is used to determine how many latest fuel blocks we should ensure are committed to l1. Current sentiment is that for the fraud proving we only need data for blocks that haven't yet been finalized. So this will probably be set to something around 600k-1M what is the current weekly block count of the testnet.

We request all blocks that fit the above description that we don't have in our database:
db

The next action happens when we either accumulate enough blocks (`blocks_to_accumulate`) or a timeout happens because we haven't submitted to L1 in a long time (`accumulation_timeout`).

We take whatever we have accumulated so far and give that to the block bundler:
bundler
The bundler will try out all possible bundles until the `optimization_timeout` runs out or we've exhausted all possibilities. 

Upon choosing a bundle candidate it will compress it and ask L1 how much gas it would cost to post it. The bundler then chooses the best candidate, the one that achieved the best gas per uncompressed byte -- meaning it posted the most amount of data for the least gas. This is a trade-off where you might include more blocks to get a better compression, but you might end up having to pay for a full extra blob that you won't utilize.

Not all combinations of blocks are permissible, a bundle must start from the lowest height and have the blocks be sorted by height. So for blocks 1,2,3,4 possible candidates for bundling are: 1,2,3,4 | 1,2,3 | 1,2 | 1. The `optimization_step` config influences how the candidates are generated. The above example was for when the step is equal to `1`. A step of `100` for a 1000 block bundle would generate the candidates whose last block has the height of: 1000, 900, ..., 100, 1, 950, 850, ..., 50, 975, 925, and so on where the step is halved upon reaching a bundle of only 1 block. 

Upon finding the best proposal or upon the `optimization_timeout` the bundle will then be given to the `L1` adapter so that it may be turned into `Fragment`s (currently blobs because we target Ethereum):
blobs

Those blobs are saved in the database where they will be picked up later by the `StateCommitter`:
committer

The committer is configured to accumulate at least `fragments_to_accumulate` blobs before sending them together in a tx. If `fragment_accumulation_timeout` fires then we will submit whatever blobs we have.

This behavior was added so that we may optimize the base eth cost and not pay it once for every blob.

All timeouts (except the optimization one) are measured from the time we last had a finalized eth transaction. This was envisioned so that if we haven't submitted for an unacceptably long time we are not going to spend time waiting for more blocks or fragments.

Finally the state listener polls for the tx status:
Screenshot_2024-09-26_14-40-33

Some notes:
* We never send a new eth tx until we confirm the last one was finalized or failed
* We work on creating new bundles even if there is congestion on the network, we use the time to optimize bundles while we wait for the blob tx to be accepted. 

Metrics:
Two new metrics were added:
```rust
    blobs_per_tx: prometheus::Histogram,
    blob_used_bytes: prometheus::Histogram,
```
which can be used to track how often we send out transactions with 1,2,3,...,6 blobs, and how utilized are those blobs.
Example query answering the question: "in 99% of cases in the last N minutes, what is the smallest amount of data we sent in a blob?"
grapy

### CODE REVIEW

File: committer/src/config.rs [hunk 1]
```diff
@@ -68,11 +74,73 @@ pub struct App {
...
+    ///
+    /// - **Bundling Behavior:**
+    ///   - **Unbundled Blocks:** Blocks outside the lookback window will not be bundled.
+    ///   - **Already Bundled Blocks:** If a block has already been bundled, its fragments will
+    ///     not be sent to L1.
+    ///   - **Failed Submissions:** If fragments of a bundled block were sent to L1 but failed,
+    ///     they will not be retried.
+    ///
+    /// This approach effectively "gives up" on blocks that fall outside the defined window.
+    pub block_height_lookback: u32,
```
Comment (digorithm) [inline committer/src/config.rs:140]: I'm still going over the code, but I have a question about this: Based on this bundling behavior, is there a chance that a block will be wrongfully excluded from bundling, thus not sending it to the L1?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -67,6 +67,9 @@ jobs:
           components: clippy,rustfmt
       - name: Install Foundry
         uses: foundry-rs/foundry-toolchain@v1
+        with:
+          # The last version of Foundry that didn't experience this issue: https://github.com/alloy-rs/alloy/issues/1371
```
Comment (digorithm) [inline .github/workflows/ci.yml:71]: We should open an issue to remind of addressing this later.

Comment (segfault-magnet) [reply .github/workflows/ci.yml:72]: https://github.com/FuelLabs/fuel-block-committer/issues/127

File: committer/src/config.rs (hunk 1 above)
Comment (segfault-magnet) [reply committer/src/config.rs:162]: This depends on how many blocks are needed for fraud proving. If we only need 1w worth of blocks we'll set this config to a value bigger than what we expect in a week. 

Otherwise if it happens we need all the blocks then we'll set this to u32::MAX and will always include all blocks from genesis onwards.

**Review by MujkicA — COMMENTED** [review]
I still have to go through some db changes and the bundle optimisation. Looks great so far, left some nits

**Review by digorithm — DISMISSED** [review]
What a journey it was reading this PR. It's pretty large, so I could be missing something; it's hard to catch bugs in PRs this large. But given our situation and deadlines and that I've been locally running this branch for a while (against the testnet) and seeing it running successfully, I think it LGTM. Nits and small improvements can be made in follow-up work. 

The tests are also incredible and helped guide this review; excellent job, @segfault-magnet. Also, thanks, @MujkicA and @hal3e, for the support on this. This is one of the coolest pieces of work I've seen lately. I'm very, very excited to see this live!

I'm approving this but I'm seconding most of @MujkicA suggestions above.

File: packages/services/src/state_committer.rs [hunk 3]
```diff
@@ -1,166 +1,423 @@
...
+        Ok(NonEmpty::collect(existing_fragments))
+    }
 
-        Ok(())
+    async fn should_submit_fragments(&self, fragment_count: NonZeroUsize) -> Result<bool> {
+        if fragment_count >= self.config.fragments_to_accumulate {
+            return Ok(true);
+        }
+        info!(
+            "have only {} out of the target {} fragments per tx per tx",
```
Comment (MujkicA) [inline packages/services/src/state_committer.rs:147]:
```suggestion
            "have only {} out of the target {} fragments per tx",
```

**Review by MujkicA — COMMENTED** [review]
Only the block bundler is left

File: packages/storage/src/mappings/tables.rs [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+            fuel_block_hash: value.block_hash.to_vec(),
+            fuel_block_height: i64::from(value.block_height),
+            completed: value.completed,
+            submittal_height: value.submittal_height.into(),
+        }
+    }
+}
+
+#[derive(sqlx::FromRow)]
+pub struct BundleFragment {
```
[the remark this answers is not in the document]
Comment (segfault-magnet) [reply packages/storage/src/mappings/tables.rs:62]: `mod`s are idiomatically used for that purpose: `committer::Config`, `watcher::Config` 

here: `tables::BundleFragment`, `tables::L1Tx`.

**Review by Br1ght0ne — APPROVED** [review]
Finished reading top to bottom. I concur, it's difficult to come up with suggestions in a PR of this size, but nothing immediately comes to mind. Might find something for a follow-up :)

edit: big props for the visual explainer too!

File: packages/eth/src/websocket/connection.rs [hunk 5]
```diff
@@ -61,13 +72,64 @@ pub struct WsConnection {
...
+            blobs_per_tx: prometheus::Histogram::with_opts(histogram_opts!(
+                "blob_per_tx",
+                "Number of blobs per blob transaction",
+                vec![1.0f64, 2., 3., 4., 5., 6.]
+            ))
+            .expect("to be correctly configured"),
+
+            blob_used_bytes: prometheus::Histogram::with_opts(histogram_opts!(
+                "blob_utilization",
+                "% utilization of blobs",
```
[the remark this answers is not in the document]
Comment (segfault-magnet) [reply packages/eth/src/websocket/connection.rs:117]: Yup, it should be "bytes filled per blob"

File: packages/eth/src/blob_encoding.rs [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+fn split_sidecar(builder: SidecarBuilder) -> crate::error::Result<NonEmpty<SingleBlob>> {
+    let num_bytes = u32::try_from(builder.len()).map_err(|_| {
+        crate::error::Error::Other("cannot handle more than u32::MAX bytes".to_string())
+    })?;
+    let sidecar = builder
+        .build()
+        .map_err(|e| crate::error::Error::Other(e.to_string()))?;
+
+    if sidecar.blobs.len() != sidecar.commitments.len()
```
[the remark this answers is not in the document]
Comment (segfault-magnet) [reply packages/eth/src/blob_encoding.rs]: Paranoia. I'll remove it, alloy seems to behave consistently wrt.

**Review by MujkicA — APPROVED** [review]

**Review by digorithm — APPROVED** [review]

**Review by hal3e — DISMISSED** [review]

**Review by Br1ght0ne — DISMISSED** [review]

**Review by hal3e — APPROVED** [review]

**Review by MujkicA — APPROVED** [review]

**Review by Salka1988 — APPROVED** [review]
